TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/RequestSenderOrchestrator.java

```diff
@@ -13,228 +14,242 @@
-        return Math.max(0, Duration.between(now(), to).toMillis());
+    private Duration getDelayFromNowMs(Instant to) {
+        return Duration.ofMillis(Math.max(0, Duration.between(now(), to).toMillis()));
     }
 
-    // TODO - rewrite this - the recursion (at least as it is) is terribly confusing
-    private void sendNextPartAndContinue(NettyPacketToHttpConsumer packetReceiver,
-                                         EventLoop eventLoop, Iterator<ByteBuf> iterator,
-                                         Instant start, Duration interval, AtomicInteger counter,
-                                         StringTrackableCompletableFuture<AggregatedRawResponse> responseFuture) {
+    private DiagnosticTrackableCompletableFuture<String,AggregatedRawResponse>
+    sendSendingRestOfPackets(IPacketFinalizingConsumer<AggregatedRawResponse> packetReceiver,
+                             EventLoop eventLoop,
+                             Iterator<ByteBuf> iterator,
+                             Instant startAt,
+                             Duration interval,
+                             AtomicInteger counter) {
         final var oldCounter = counter.getAndIncrement();
         log.atTrace().setMessage(()->"sendNextPartAndContinue: counter=" + oldCounter).log();
         assert iterator.hasNext() : "Should not have called this with no items to send";
 
-        packetReceiver.consumeBytes(iterator.next());
+        var consumeFuture = packetReceiver.consumeBytes(iterator.next());
         if (iterator.hasNext()) {
-            Runnable packetSender = () -> sendNextPartAndContinue(packetReceiver, eventLoop,
-                    iterator, start, interval, counter, responseFuture);
-            var delayMs = Duration.between(now(),
-                    start.plus(interval.multipliedBy(counter.get()))).toMillis();
-            eventLoop.schedule(packetSender, Math.min(0, delayMs), TimeUnit.MILLISECONDS);
+            return consumeFuture.thenCompose(cf ->
+                    NettyToCompletableFutureBinders.bindNettyScheduleToCompletableFuture(eventLoop,
+                                    Duration.between(now(), startAt.plus(interval.multipliedBy(counter.get()))))
+                    .getDeferredFutureThroughHandle((v,t)-> sendSendingRestOfPackets(packetReceiver, eventLoop,
+                                iterator, startAt, interval, counter), () -> "sending next packet"),
+                            () -> "recursing, once ready");
         } else {
-            packetReceiver.finalizeRequest().handle((v,t)-> {
-                if (t != null) {
-                    responseFuture.future.completeExceptionally(t);
-                } else {
-                    responseFuture.future.complete(v);
-                }
-                return null;
-            }, ()->"waiting for finalize to send Aggregated Response");
+            return consumeFuture.getDeferredFutureThroughHandle((v,t) -> packetReceiver.finalizeRequest(),
```

Comment:
It hasn't been a concern for me because finalizeRequest() will itself be called from a chained future from consumeBytes.  The consumeBytes Future itself will internally be used by finalizeRequest to chain finalize.  finalizeRequest() will handle exceptions by setting the AggregateRawResponse error cause.  That's then sent through to the tuple log.

We might want more, but we didn't have it  yet.  I suspect that once retries come in, we'll probably need some more clues.